inst/include/common/model.hpp

```diff
@@ -86,99 +84,171 @@ class Model {  // may need singleton
     vector<vector<Type> > log_recruit_dev(n_pops);
     vector<vector<Type> > recruitment(n_pops);
     vector<vector<Type> > M(n_pops);
-
+    vector<Type> nll_components(this->fims_information->density_components.size());
 #endif
+    // Loop over densities and evaluate joint negative log densities for priors
+    typename fims_info::Information<Type>::density_components_iterator d_it;
+    nll_components.fill(0);
+    int nll_components_idx = 0;
+    size_t n_priors = 0;
+    MODEL_LOG << "Expecting to evaluate " << this->fims_information->density_components.size()
+               << " density components." << std::endl;
+    for(d_it = this->fims_information->density_components.begin();
```

Comment:
I wonder if the for loop can be generalized for different d->input_type (lines 96-111, 137-158, and 194-210) and converted into one helper function. This may reduce code duplication.

Reply:
Thanks for the suggestion! I'll add that to issue #613.